proposals/p2029.md

```diff
@@ -0,0 +1,103 @@
+# C++ Interop: Reserved identifiers
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2029)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+C++ reserves certain sets of identifiers for implementation-specific use. Carbon
+must ensure it respects these reserved identifiers to enable seamless and
+error-free interoperability with C++. Equally, it would be prudent to define a
+set of identifiers which are reserved for internal Carbon implementation usage.
+
+## Background
```

Comment:
What about C++ keywords?

Reply:
In the Background section, I link  to the (stalled) PR which discusses this: https://github.com/carbon-language/carbon-lang/pull/93

Reply:
I think Carbon will have some syntax like Rust's `r#` "raw identifier" syntax for saying "treat this Carbon keyword as an identifier", as discussed in #93 . This will allow C++ APIs using Carbon keywords to be usable from Carbon. The bigger problem is using Carbon APIs that use C++ keywords from C++.

This problem might be out of scope for this proposal, but it seems very related (not using things in Carbon that aren't legal in C++), and worth mentioning. I wonder if there might be a single solution that works for things with underscores as well as C++ keywords. Possibly an "export to C++ as X" annotation?

Reply:
I can happily throw in a section on this. I think there are a number of options for the keyword issue:

1. Do nothing - users will need to refactor their code if they hit problems when using their libraries in C++
2. Do nothing in regular compilation, error / warn if emitting a C++ header
3. Always error or warn (which is an option presented in #93)
4. Allow an "export to C++ as X" annotation

I note that option 4 can be combined with 1-3 to provide a route to fix which doesn't change the whole Carbon codebase too.

I've seen some discussion flying around on what annotations could look like, but not sure what the current 'state of the art' is; is there a working syntax for now?